File: python/05_query_soql.py
"""Step 5: SOQL queries using simple-salesforce.

Demonstrates sf.query() (up to 2000 records) and sf.query_all() (full pagination),
plus relationship queries across objects.

Run with:
    uv run python 05_query_soql.py
"""

from auth import get_salesforce_connection


def query_accounts(sf) -> None:
    print("--- Query: Accounts (LIMIT 10) ---")
    result = sf.query("SELECT Id, Name, Phone, Industry FROM Account LIMIT 10")
    print(f"  Records returned: {len(result['records'])} of {result['totalSize']} total")
    for record in result["records"]:
        print(f"  {record['Id']} | {record['Name']} | {record.get('Industry') or 'N/A'}")


def query_contacts_with_account(sf) -> None:
    print("--- Query: Contacts with Account name (relationship query) ---")
    result = sf.query(
        "SELECT Id, FirstName, LastName, Email, Account.Name FROM Contact LIMIT 10"
    )
    print(f"  Records returned: {len(result['records'])}")
    for record in result["records"]:
        account_name = record.get("Account", {}) or {}
        account_name = account_name.get("Name") if account_name else "No Account"
        name = f"{record.get('FirstName') or ''} {record['LastName']}".strip()
        print(f"  {name} | {record.get('Email') or 'N/A'} | Account: {account_name}")


def query_open_cases(sf) -> None:
    print("--- Query: Open Cases ---")
    result = sf.query(
        "SELECT Id, Subject, Status, Priority, Account.Name FROM Case "
        "WHERE Status != 'Closed' LIMIT 10"
    )
    print(f"  Open cases: {len(result['records'])}")
    for record in result["records"]:
        account_name = record.get("Account", {}) or {}
        account_name = account_name.get("Name") if account_name else "No Account"
        print(f"  [{record['Priority']}] {record['Subject']} | {record['Status']} | {account_name}")


def query_all_accounts(sf) -> None:
    print("--- Query All: Accounts (handles pagination automatically) ---")
    # query_all() follows nextRecordsUrl automatically until all records are retrieved.
    # Use this when you expect more than 2,000 records.
    result = sf.query_all("SELECT Id, Name FROM Account")
    print(f"  Total accounts in org: {result['totalSize']}")
    print(f"  Records retrieved: {len(result['records'])}")


def main():
    sf = get_salesforce_connection()

    query_accounts(sf)
    print()
    query_contacts_with_account(sf)
    print()
    query_open_cases(sf)
    print()
    query_all_accounts(sf)

    print("\nDone.")


if __name__ == "__main__":
    main()

File: javascript/05_query_soql.js
'use strict';
/**
 * Step 5: SOQL queries using jsforce.
 *
 * Demonstrates conn.query() (up to 2000 records) and conn.queryAll() (full pagination),
 * plus relationship queries across objects.
 *
 * Run with:
 *   node 05_query_soql.js
 */

const { getConnection } = require('./auth');

async function queryAccounts(conn) {
    console.log('--- Query: Accounts (LIMIT 10) ---');
    const result = await conn.query(
        'SELECT Id, Name, Phone, Industry FROM Account LIMIT 10'
    );
    console.log(`  Records returned: ${result.records.length} of ${result.totalSize} total`);
    result.records.forEach(r => {
        console.log(`  ${r.Id} | ${r.Name} | ${r.Industry || 'N/A'}`);
    });
}

async function queryContactsWithAccount(conn) {
    console.log('--- Query: Contacts with Account name (relationship query) ---');
    const result = await conn.query(
        'SELECT Id, FirstName, LastName, Email, Account.Name FROM Contact LIMIT 10'
    );
    console.log(`  Records returned: ${result.records.length}`);
    result.records.forEach(r => {
        const name = [r.FirstName, r.LastName].filter(Boolean).join(' ');
        const accountName = r.Account ? r.Account.Name : 'No Account';
        console.log(`  ${name} | ${r.Email || 'N/A'} | Account: ${accountName}`);
    });
}

async function queryOpenCases(conn) {
    console.log('--- Query: Open Cases ---');
    const result = await conn.query(
        "SELECT Id, Subject, Status, Priority, Account.Name FROM Case WHERE Status != 'Closed' LIMIT 10"
    );
    console.log(`  Open cases: ${result.records.length}`);
    result.records.forEach(r => {
        const accountName = r.Account ? r.Account.Name : 'No Account';
        console.log(`  [${r.Priority}] ${r.Subject} | ${r.Status} | ${accountName}`);
    });
}

async function queryAllAccounts(conn) {
    console.log('--- Query All: Accounts (handles pagination automatically) ---');
    // queryAll() follows nextRecordsUrl automatically until all records are retrieved.
    // Use this when you expect more than 2,000 records.
    const result = await conn.queryAll('SELECT Id, Name FROM Account');
    console.log(`  Total accounts in org: ${result.totalSize}`);
    console.log(`  Records retrieved: ${result.records.length}`);
}

async function main() {
    const conn = await getConnection();

    await queryAccounts(conn);
    console.log();
    await queryContactsWithAccount(conn);
    console.log();
    await queryOpenCases(conn);
    console.log();
    await queryAllAccounts(conn);

    console.log('\nDone.');
}

main().catch(err => {
    console.error('Error:', err.message);
    process.exit(1);
});
